### PR TITLE
Fixes RCA crash on active master (#131)

### PR DESCRIPTION
**Pull request by sgup432**
Signed-off-by: Sagar Upadhyaya <sagar.upadhyaya.121@gmail.com>

**Is your feature request related to a problem? Please provide an existing Issue # , or describe.**
https://github.com/opensearch-project/performance-analyzer-rca/issues/131

**Describe the solution you are proposing**
- Keeps track of which streams were already completed by client via a flag "isCompleted". When RCA is disabled, publisher(active master) checks whether this response stream was already completed. If it was completed, it doesn't invoke onNext on response stream to convey it is shutting down.

**Describe alternatives you've considered**

**Additional context**

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/performance-analyzer-rca/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by sruti1312 — DISMISSED** [review]

File: src/main/java/org/opensearch/performanceanalyzer/rca/net/handler/PublishRequestHandler.java [hunk 1]
```diff
@@ -68,28 +68,49 @@ public PublishRequestHandler(
...
+    // Visible for testing
+    protected List<SendDataClientStreamUpdateConsumer> getDataClientStreamList() {
+        return this.dataClientStreamList;
     }
 
-    private class SendDataClientStreamUpdateConsumer implements StreamObserver<FlowUnitMessage> {
+    protected class SendDataClientStreamUpdateConsumer implements StreamObserver<FlowUnitMessage> {
 
         private final StreamObserver<PublishResponse> serviceResponse;
+        private boolean isCompleted;
```
Comment (khushbr) [inline src/main/java/org/opensearch/performanceanalyzer/rca/net/handler/PublishRequestHandler.java:104]: Let's make the variable `final`. 

Also, would recommend renaming it to `isClosed`.

Comment (sgup432) [reply src/main/java/org/opensearch/performanceanalyzer/rca/net/handler/PublishRequestHandler.java:104]: Actually can't make it final as will then have to initialize it to a default false value. And then I won't be able to change its value later on as I am doing below.
I used isCompleted as per grpc stream terminology where we usually "complete" the response stream in the end. Though it is similar to close.

Let me know if you still think that we should rename it.

Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/rca/net/handler/PublishRequestHandler.java:104]: Makes sense, we can keep the naming. 

I wanted to confirm one more thing, is the `isCompleted` variable value updated only by single owner thread ? Do we need `AtomicBoolean` here?

Comment (sgup432) [reply src/main/java/org/opensearch/performanceanalyzer/rca/net/handler/PublishRequestHandler.java:104]: Don't need a AtomicBoolean here as this object is created per node per vertex.

Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/rca/net/handler/PublishRequestHandler.java:104]: LGTM then.

**Review by khushbr — APPROVED** [review]

**Pull request merged**